pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/memory/unsafe/DirectMemory.java

```diff
@@ -55,23 +54,18 @@ public void flush() {
   }
 
   @Override
-  public void close()
-      throws IOException {
+  public synchronized void close() {
     if (!_closed) {
-      synchronized (this) {
-        if (!_closed) {
-          Unsafer.UNSAFE.freeMemory(_address);
-          _closed = true;
-        }
-      }
+      Unsafer.UNSAFE.freeMemory(_address);
```

Comment:
I think we have to recover the synchronization here. Otherwise a race condition may try to free an address that was already cleaned. That could end up producing a SIGSEV which would kill the server.

We can improve the code using double check

Reply:
The whole method is changed to `synchronized`. I don't remember where I read it, but for modern JVM double checking synchronization is not necessary, especially when we almost always only call it once.

Reply:
You are right, I didn't see that. Anyway, in the other PR I've changed that so we don't need to synchronize in case one thread close this after another.

What newer jvms does for a fact is to make synchronized by default is specially expensive. This is due to:
* Virtual threads being pinned with synchronize. This is secondary given we don't use them and it should be something that eventually will be improved in (even) more modern JVMs
* +UseBiasedLocking has been disabled since Java 15 and the plan is to remove it. See https://openjdk.org/jeps/374, [this](https://dev.to/vbochenin/java-17-migration-bias-locks-regression-2c5m#:~:text=What%20is%20a%20bias%20lock,the%20same%20thread%20is%20lightweight.) and [this](https://quarkus.io/blog/biased-locking-help/)

Reply:
Why would multiple threads trying to close the same buffer? If that happens, that means the buffer ownership is messed up. In regular case, this method should only be invoked exactly once.
After reading [this](https://www.cs.umd.edu/~pugh/java/memoryModel/DoubleCheckedLocking.html), I'm even more worried about double-checked locking..

Reply:
Ok, lets close the other PR then.